### PR TITLE
Fix winston logging

### PR DESCRIPTION
**Pull request by somewhatabstract**
This update fixes our use of winston by ensuring we create a logger configured with the appropriate transports. It also adds some formatters to support the %s substitutions that we use so that messages properly populate. Finally, we replace our custom profile approach with the winston variant and a formatter that includes the duration information in the output message.

## Additional Changes
This also makes it so that logging during tests goes to a stream rather than the console to keep things quiet.

Logging initialization occurs in a new `logging.js` file.

Argument parsing has been moved to a separate file so we can reuse it in the new `logging.js` file. The parsing will only happen once. If it is loaded outside of the RRS execution (such as by testing), then it returns some default args instead for running in the test environment.

## Testing
To test, I did the following:

**tests**
`npm test` and verified things pass and that the output is as we would expect

**dev/prod**
Modified webapp to use this variant locally and ran `make serve` to see that it was working (also looked at the react-render-server.log file).

Ran it to be with or without --dev just to check things run.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
     "debug": "nodemon -i src/render_cache src/main.js -- --dev --log-level=debug",
     "profile": "node-debug -p 8040 --no-preload src/main.js --dev",
     "serve_local": "nodemon -i src/render_cache src/main.js -- --dev",
-    "test": "mocha --exit --reporter spec 'src/*_test.js'",
+    "test": "NODE_ENV=test mocha --exit --reporter spec 'src/*_test.js'",
```
Comment (somewhatabstract) [inline package.json]: `mocha` doesn't set `NODE_ENV` to `test`, so we do it here instead.

File: src/arguments.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        defaultValue: 1000,
+        help: "How many ms until we abort a render as taking too long.",
+    });
+parser.addArgument(
+    ['--log-level'],
+    {
+        defaultValue: 'info',
+        choices: ['silly', 'debug', 'verbose', 'info', 'warn', 'error'],
+        help: "What level to log at.",
+    });
```
Comment (somewhatabstract) [inline src/arguments.js]: All copypasta from `main.js` except...

File: src/arguments.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+// We only want to parse the args if we're running inside our main app.
+const args = process.argv[1].endsWith("/src/main.js")
+    ? parser.parseArgs()
+    : {
+        // Some defaults for tests and the like.
+        log_level: "debug",
+        dev: true,
+    };
```
Comment (somewhatabstract) [inline src/arguments.js]: These defaults when the running file isn't `main.js`

File: src/create-render-context.js [hunk 4]
```diff
@@ -229,11 +229,11 @@ const createRenderContextWithStats = function(
     locationUrl,
     globals,
     jsPackages,
-    pathToClientEntryPoint,
     requestStats,
 ) {
-    const vmConstructionProfile = profile.start("building VM for " +
-        pathToClientEntryPoint);
```
Comment (somewhatabstract) [inline src/create-render-context.js]: Discovered this was always undefined so I modified it to be more useful

File: src/logging.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
+/**
```
Comment (somewhatabstract) [inline src/logging.js]: This is part new work and part copypasta from `main.js`. The new work is to configure winston itself and to add our formatters.

File: src/main.js [hunk 6]
```diff
@@ -88,33 +45,14 @@ if (args.dev) {
...
-}));
+appWithLogging.use(logging.middleware.requestLogger);
+appWithLogging.use(logging.middleware.errorLogger);
 appWithLogging.use(app);
 
 const server = appWithLogging.listen(port, () => {
     const host = server.address().address;
     const port = server.address().port;
-    logging.info('react-render-server running at http://%s:%s', host, port);
+    logging.info(`react-render-server running at http://%s:%s`, host, port);
```
Comment (somewhatabstract) [inline src/main.js]: I had used interpolation to fix this message until I discovered the `splat()` formatter that allows for the %s approach. I must've forgotten to put the quotes back the way they were.

File: src/render.js [hunk 7]
```diff
@@ -80,7 +80,7 @@ const performRender = async () => {
  * That entrypoint should call __registerForSSR__ which gives us a callback
  * to actually get the rendered content.
  *
- * @param {[{url: string, content: string}]} jsPackages - Absolute URL to the
+ * @param {[Script]} jsPackages - Absolute URL to the
```
Comment (somewhatabstract) [inline src/render.js]: It was wrong. Now it's not (at least until I change things again).

File: src/render.js [hunk 8]
```diff
@@ -108,20 +108,19 @@ const render = async function(
     globals,
     requestStats,
 ) {
-    const {url: entryPointUrl} = jsPackages[jsPackages.length - 1];
```
Comment (somewhatabstract) [inline src/render.js]: There is no `url` property. This was aways setting `entryPointUrl` to `undefined`. Luckily for us, it was only used in logging.

File: src/server.js [hunk 9]
```diff
@@ -87,7 +87,7 @@ app.use("/render", (req, res, next) => {
     };
 
     pendingRenderRequests++;
-    const renderProfile = profile.start();
+    const renderProfile = profile.start("/render");
```
Comment (somewhatabstract) [inline src/server.js]: I made it so profile calls had to provide a message.

File: src/server_test.js [hunk 10]
```diff
@@ -219,17 +223,17 @@ describe("API endpoint /render", function() {
...
                     foundEntry = true;
-                    matchedStats = arg;
+                    matchedStats = message;
                 }
             });
         });
         assert.isTrue(
             foundEntry,
-            `No stats entry like ${expectedEntry}. ${JSON.stringify(debugLoggingSpy.args)}`,
+            `No stats entry like ${expectedEntry}.\n\n${JSON.stringify(doneFake.args)}`,
```
Comment (somewhatabstract) [inline src/server_test.js]: Changing how the profiling worked meant fixing this test case.

**Comment by csilvers** [comment]
I don't suppose you could do these code reviews via phabricator?  That would make life easier for me, at least.

I find it really difficult to do code reviews in github, so I haven't looked closely at this diff.  I would suggest, though, that you look at services/graphql-gateway/src/main.js and see how it does logging.  It would be good to normalize on the same approach now, so that when we move the render-server into webapp, they can share the same logging code.

The graphql-gateway logging is, in turn, based on the react-render-server, though maybe I was using the right transport from the beginning.  It's probably simpler.  As long as what you have is a strict superset of what graphql-gateway does, that's fine, we can use your version when you go into webapp.  If they differ, though, we should figure out why now, and make sure they match.

File: src/profile.js [hunk 11]
```diff
@@ -11,29 +13,30 @@
...
         end: (endMsg) => {
-            const endTime = process.hrtime();
-            logging.debug('PROFILE: %s (finished after %d ms)',
-                endMsg || msg,
-                (hrtToMs(endTime) - hrtToMs(startTime)).toFixed(4));
+            const message = endMsg || msg;
+            profiler.done({
+                message: `PROFILE(end): ${message}`,
+                level: "debug",
+            });
```
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply src/profile.js]: The reason I ask it that it looked like we were doing some calculations before to determine the duration of different things.  I wanted to confirm that's no longer necessary.

[the remark this answers is not in the document]
Comment (somewhatabstract) [reply src/profile.js]: I rewrote the profiling to not have to do our own calculations.

**Pull request merged**